icu4c/source/i18n/unitconverter.cpp

```diff
@@ -5,14 +5,74 @@
 
 #if !UCONFIG_NO_FORMATTING
 
+#include "charstr.h"
+#include "measunit_impl.h"
+#include "unicode/errorcode.h"
+#include "unicode/measunit.h"
+#include "unicode/stringpiece.h"
+#include "unitconverter.h"
+
 U_NAMESPACE_BEGIN
 
+namespace {
+
+const ConversionRateInfo &
+extractConversionInfo(StringPiece source,
+                      const MaybeStackVector<ConversionRateInfo> &conversionRateInfoList,
+                      UErrorCode &status) {
+    for (size_t i = 0, n = conversionRateInfoList.length(); i < n; ++i) {
+        if (conversionRateInfoList[i]->sourceUnit.toStringPiece() == source)
+            return *(conversionRateInfoList[i]);
+    }
+
+    status = U_INTERNAL_PROGRAM_ERROR;
+    return ConversionRateInfo();
+}
+
+MeasureUnit extractBaseUnit(const MeasureUnit &source,
+                            const MaybeStackVector<ConversionRateInfo> &conversionRateInfoList,
+                            UErrorCode &status) {
+    MeasureUnit result;
+    int32_t count;
+    const auto singleUnits = source.splitToSingleUnits(count, status);
+    if (U_FAILURE(status)) return result;
+
+    for (int i = 0; i < count; ++i) {
+        const auto &singleUnit = singleUnits[i];
+        // Extract `ConversionRateInfo` using the absolute unit. For example: in case of `square-meter`,
+        // we will use `meter`
+        const auto singleUnitImpl = SingleUnitImpl::forMeasureUnit(singleUnit, status);
+        const auto &rateInfo =
+            extractConversionInfo(singleUnitImpl.identifier, conversionRateInfoList, status);
+        if (U_FAILURE(status)) return result;
+
+        // Multiply the power of the singleUnit by the power of the baseUnit. For example, square-hectare
+        // must be p4-meter. (NOTE: hectare --> square-meter)
+        auto baseUnit = MeasureUnit::forIdentifier(rateInfo.baseUnit.toStringPiece(), status);
+        auto singleBaseUnit = SingleUnitImpl::forMeasureUnit(baseUnit, status);
```

Comment:
This line then fails. (e.g. if you're trying to convert "volts".)

What I assumed when I was testing this kind of code before, is that any unit that has this kind of base unit will typically not appear with a higher power (e.g. "square-volt" doesn't happen in practice). Consequently there's no need to raise baseUnit to a higher power.

I suggest you don't create a SingleUnitImpl instance here, just stick with baseUnit. There's a withDimensionality() function that you can call as long as baseUnit.getComplexity returns UMEASURE_UNIT_SINGLE.

Reply:
https://github.com/younies/icu/pull/26 fixes the problem,
https://github.com/younies/icu/pull/27 contains the unit tests that demonstrate the problem (fixed by #26).

Reply:
If the source unit has a power, the power should be distributed to all of the single units in the base unit (rules of algebra).  The easiest way to do this is to perform that operation via a MeasureUnitImpl, where you can loop over the constituent SimpleUnitImpls and set their dimensionality directly.

Reply:
Understood. Not yet implemented.

I've now pulled the unitsTest.txt cases that were failing into this pull request:
https://github.com/younies/icu/pull/28

And this pull request fixes these unit tests:
https://github.com/younies/icu/pull/29
- but does not yet fix the general case (e.g. doesn't deal with square-volt). I think this general case won't show up, but I suppose we might as well fix that too.

@younies let me know if you'd like to fix it further, or whether you'd be happy if I fixed #29 further. Up to you. I just didn't want to provide #28 breaking your unit tests without also providing a fix (e.g. #29) - I needed to write a fix anyway to more easily make sure we don't still have blocking bugs in the lower-level libraries we depend on.

Reply:
Thanks for fixing! Marking as resolved.